doc/man/man1/logic_analyzer.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" (C) Copyright 2013 Yves Renier <yrenier@cern.ch>,
.\"
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH LOGIC-ANALYZER-RPI 1 "March 22, 2013"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
logic\-analyzer\-rpi \- program to aquire logic states at the GPIO pins of a Raspberry Pi
.SH SYNOPSIS
\fIlogic\-analyzer\-rpi\fR [\-h] [\-v level] [\-w sec] [\-d protocol]  <[\-l file] | [\-c pins [\-i ms] [\-n npoint]]> <[\-s file]|[\-p file]>
.br
.SH DESCRIPTION
This manual page documents briefly the
.B logic-analyzer-rpi
and
.B bar
commands.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics,
.\" respectively.
\fBlogic-analyzer-rpi\fP is a program that can capture logic states at the 
Raspberry Pi GPIO's pins, save/load capture files, decode according to included
protocols and plot results.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-v, \-\-verbose  level
Set the verbosity to level (0 to 4, higher means more messages,default is 3).
.TP
.B \-w, \-\-wait  sec
Wait the specified number of seconds before capturing.
.TP
.B \-d, \-\-decode  protocol
Decode according to protocol. available protocols are :
.br
.RI microwire
.br
.RI raw
.TP
.B \-l, \-\-load  file
Load file instead of capturing.
.TP
.B \-c, \-\-capture  pins
Capture using given pins. Pins must be the GPIO numbers separated by \',\' or \' \'. Space may be used if le string pins is protected by \"\". ex 14,15,27,4 or \"14 15 27 4\". Order may be inportant if --decode is used depending on the protocol.
.TP
.B \-i, \-\-interval  ms
Interval between 2 logic state capture in milliseconds.
.TP
.B \-n, \-\-npoint  npoint
Number of points to capture.
.TP
.B \-s, \-\-save  file
Save capture to file.
.TP
.B \-p, \-\-plot  file
Plot into file (PNG image).
.SH SEE ALSO
.BR activity (3),
.BR binary (3),
.BR logic_input (3),
.BR message (3),
.BR microwire (3),
.BR protocol (3),
.BR raw (3),
.BR transition (3).
.br



src/logic_analyzer.cpp
/*  Copyright (C) 2012,2013 Renier Yves
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <mgl/mgl_zb.h>
#include <unistd.h>
#include <stdio.h>
#include <getopt.h>

#ifndef RAW
#define RAW
#include "raw.h"
#endif

#ifndef MICROWIRE
#define MICROWIRE
#include "microwire.h"
#endif

debug_level msglevel=INFORMATION;
//int msglevel=DEBUG;

void usage(char *program_name)
{
	printf("usage: %s [-h] [-v level] [-w sec] [-l file | -c pins [-i ms] [-n npoint]] [-d protocol] [-s file] [-p file].\n",program_name);
	printf("program displaying GPIO states over time and decoding protocols if supported.\nprotocols supported are :\n- microwire\n- raw\nprogram parameters are:\n");
	printf("\t -h|--help : this help\n");
	printf("\t -l|--load file : Use data in binary file as captured data.\n");
	printf("\t -c|--capture pins: pins to use durring the capture. Must be the GPIO numbers separated by \',\' or \' \'. Space may be used if le string pins is protected by \"\". ex 14,15,27,4 or \"14 15 27 4\". Order may be inportant if --decode is used depending on the protocol.\n");
	printf("\t -i|--interval ms: interval in ms between 2 captures (default is 0.5 ms wich is the recomended minimum). For low intervals (ms<10), the interval can be sometime higher if the RPi is busy durring capture.\n");
	printf("\t -n|--npoint npoint: number of data point to capture (default 2000).\n");
	printf("\t -d|--decode protocol: decode the captured data using the specified protocol. Some protocols may need a special order for the pins given to the --capture option.\n");	
	printf("\t -s|--save file : save captured data to binary file.\n");
	printf("\t -p|--plot file : plot captured data to PNG file.\n");
	printf("\t -w|--wait sec : time in second to wait before the capture.\n");
	printf("\t -v|--verbose level: set the verbosity to level (0 to 4, higher means more messages,default is 3).\n");
}


int main(int argc, char **argv)
{
	opterr = 0;
    int option_index = 0;
	float wait = 0;
	float interval =0.5;
	uint8_t pins[256];
	uint8_t npin=0;
	uint32_t npoint = 0;
	char loadfile[256] = "";
	char savefile[256] = "";
	char plotfile[256] = "";
	char protocol_name[25] = "";
	int c;
	static int save_flag=0,wait_flag=0,help_flag=0,plot_flag=0;
	static int load_flag=0,capture_flag=0,decode_flag=0;
	char prog_name[256] = "";
    protocol* capture=NULL;


	strncpy(prog_name,argv[0],256);
 	while (1)
	{
		static struct option long_options[] =
		{
			{"save", required_argument,  0, 's'},
			{"wait", required_argument, 0, 'w'},
			{"verbose", required_argument, 0, 'v'},
			{"help", no_argument, 0, 'h'},
			{"load", required_argument, 0, 'l'},
			{"capture", required_argument, 0, 'c'},
			{"interval", required_argument, 0, 'i'},
			{"npoint", required_argument, 0, 'n'},
			{"decode", required_argument,  0, 'd'},
			{"plot", required_argument,  0, 'p'},
			{0, 0, 0, 0}
		};
		/* getopt_long stores the option index here. */
	 
		c = getopt_long (argc, argv, "hv:w:l:c:i:n:d:s:p:",long_options, &option_index);
	 
		if (c == -1)
			break;

		switch (c)
		{
			case 'h': //help
				help_flag = 1;
				break;
			case 'v': //verbose
				msglevel=(debug_level) atoi(optarg);
				break;
			case 'w': //wait
				wait_flag = 1;
				wait=atof(optarg);
				break;
			case 'l': //load
				load_flag = 1;
				strncpy(loadfile,optarg,256);
				break;
			case 'c': //capture
				capture_flag = 1;
				npin=0;
				for (const char* p = strtok(optarg, ", "); p ;p=strtok(NULL, ", "))
				{
				    pins[npin]=atoi(p);
				    npin++;
				    if(npin>=256)
				        break;
				}
				break;
			case 'i': //interval
				interval = atof(optarg);
				break;
			case 'n': //npoint
				npoint= atoi(optarg);
				break;
			case 'd': //decode
				decode_flag=1;
				strncpy(protocol_name,optarg,25);
				break;
			case 's': //save
				save_flag=1;
				strncpy(savefile,optarg,256);
				break;
			case 'p': //plot
				plot_flag=1;
				strncpy(plotfile,optarg,256);
				break;
			case '?': 
			    /*getopt_long already printed an error message*/
				break;
			default:
				usage(prog_name);
				return 1;
		}
	}
	for (; optind < argc; optind++)
	{
		pmesg(WARNING,"Non-option argument %s\n", argv[optind]);
    }
	if(help_flag)
	{
		usage(prog_name);
		return 1;
	}

//Check for incompatibilities in paresed options
	if(!load_flag && !capture_flag)
	{
		usage(prog_name);
		pmesg(CRITICAL,"%s: no action given (--load or --capture needed).\n",prog_name);
		return 1;
	}
	if(load_flag && capture_flag)
	{
		usage(prog_name);
		pmesg(CRITICAL,"--load and --capture can t be used at the same time\n");
		return 1;
	}
//do what has been required	
    if(wait_flag)
    {
        if(wait>=1)
            bcm2835_delay(wait);
        if(wait<1 && wait>=0.001)
            bcm2835_delayMicroseconds(wait*1000);        
    }
    if( !decode_flag )
    {
        pmesg(INFORMATION,"no decoding, creating raw object...\n");
        capture=new raw();
    }
    else if(strcmp(protocol_name,"raw")==0)
    {
        pmesg(INFORMATION,"%s decoding, creating raw object...\n",protocol_name);
        capture=new raw();
    }
    else if(strcmp(protocol_name,"microwire")==0)
    {
        pmesg(INFORMATION,"%s decoding, creating microwire object...\n",protocol_name);
        capture=new microwire();
    }
    else
    {
        usage(prog_name);
	    pmesg(CRITICAL,"unknown protocol \"%s\".\n",protocol_name);
	    return 1;
	}
    
    if(capture_flag)
    {
        pmesg(INFORMATION,"init acquisition...\n");
        capture->init_acquisition(pins,npin);
        pmesg(INFORMATION,"acquire...\n");
        capture->acquire(npoint,interval);
    }
    if(load_flag)
    {
        pmesg(INFORMATION,"Load %s...\n",loadfile);
        capture->Load(loadfile);
    }
    if(save_flag)
    {
        pmesg(INFORMATION,"Save in %s...\n",savefile);
        capture->Save(savefile);
    }
    if(decode_flag)
    {
        pmesg(INFORMATION,"decode...\n");
        capture->decode();
    }
    if(plot_flag)
    {
        pmesg(INFORMATION,"create graph window...\n");
        mglGraphZB gr(1280,1024);
        pmesg(INFORMATION,"plot...\n");
        capture->Draw(&gr);
        pmesg(INFORMATION,"Write PNG\n");
        gr.WritePNG(plotfile);
    }
    if(capture!=NULL)
        delete capture;
    return 0;
}
